### PR TITLE
FRAME: Consideration (abstraction over storage deposits)

### PR DESCRIPTION
**Pull request by gavofyork**
One half of paritytech/substrate#13666

Changes to Preimage pallet to utilize this will be in the [follow-up PR](https://github.com/paritytech/polkadot-sdk/pull/1363).

This doesn't touch any live logic, so can go in without audit once reviewed.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3536065

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3537104

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3537238

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3537984

File: substrate/frame/support/src/traits/storage.rs [hunk 1]
```diff
@@ -127,6 +132,69 @@ impl WhitelistedStorageKeys for Tuple {
...
+	/// For creating tickets and dropping them, you can use the simpler `new` and `drop` instead.
+	fn update(
+		who: &AccountId,
+		old: Option<Self::Ticket>,
+		new: Option<Footprint>,
+	) -> Result<Self::Ticket, DispatchError>;
+
+	/// Create a ticket for the `new` footprint attributable to `who`. This ticket *must* be
+	/// consumed (through either `drop` or `update`) if the footprint changes or is removed.
+	fn new(who: &AccountId, new: Footprint) -> Result<Self::Ticket, DispatchError> {
```
Comment (KiChjang) [inline substrate/frame/support/src/traits/storage.rs:181]: Looks like this function would benefit from the [`#[must_use]` attribute](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute)?

File: substrate/frame/support/src/traits/tokens/fungible/mod.rs [hunk 2]
```diff
@@ -55,3 +61,179 @@ pub use item_of::ItemOf;
...
+	fn update(self, who: &A, footprint: Footprint) -> Result<Self, DispatchError> {
+		F::set_on_hold(&R::get(), who, D::convert(footprint)).map(|_| Self(PhantomData))
+	}
+	fn drop(self, who: &A) -> Result<(), DispatchError> {
+		F::release_all(&R::get(), who, BestEffort).map(|_| ())
+	}
+}
+
+#[test]
+fn it_builds() {}
```
Comment (ggwpez) [inline substrate/frame/support/src/traits/tokens/fungible/mod.rs:239]:
```suggestion
```

**Review by ggwpez — APPROVED** [review]

File: substrate/frame/support/src/traits/storage.rs [hunk 3]
```diff
@@ -127,6 +132,66 @@ impl WhitelistedStorageKeys for Tuple {
...
+		Self { count: items as u64, size: len as u64 }
+	}
+
+	pub fn from_encodable(e: impl Encode) -> Self {
+		Self::from_parts(1, e.encoded_size())
+	}
+}
+
+/// Some sort of cost taken from account temporarily in order to offset the cost to the chain of
+/// holding some data `Footprint` in state.
```
Comment (kianenigma) [inline substrate/frame/support/src/traits/storage.rs:156]:
```suggestion
/// holding some data [`Footprint`] in state.
```

for types in the local file this is a pretty low effort way to make the rust-docs link-able, but it won't work as easily for items that are not in the current file.

**Review by kianenigma — APPROVED** [review]

File: substrate/frame/support/src/traits/tokens/fungible/freeze.rs [hunk 4]
```diff
@@ -65,4 +71,53 @@ pub trait Mutate<AccountId>: Inspect<AccountId> {
...
+		who: &AccountId,
+		amount: Self::Balance,
+		fortitude: Fortitude,
+	) -> DispatchResult {
+		let force = fortitude == Fortitude::Force;
+		ensure!(force || Self::balance_freezable(who) >= amount, TokenError::FundsUnavailable);
+		Self::extend_freeze(id, who, amount)
+	}
+
+	/// Decrease the amount which is being frozen for a particular lock, failing in the case of
```
Comment (franciscoaguirre) [inline substrate/frame/support/src/traits/tokens/fungible/freeze.rs:106]:
```suggestion
	/// Decrease the amount which is being frozen for a particular freeze, failing in the case of
```
?

File: substrate/frame/support/src/traits/tokens/fungible/freeze.rs [hunk 5]
```diff
@@ -65,4 +71,53 @@ pub trait Mutate<AccountId>: Inspect<AccountId> {
...
+	/// underflow.
+	fn decrease_frozen(id: &Self::Id, who: &AccountId, amount: Self::Balance) -> DispatchResult {
+		let a = Self::balance_frozen(id, who)
+			.checked_sub(&amount)
+			.ok_or(ArithmeticError::Underflow)?;
+		Self::set_freeze(id, who, a)
+	}
+
+	/// Increase the amount which is being frozen for a particular lock, failing in the case that
```
Comment (franciscoaguirre) [inline substrate/frame/support/src/traits/tokens/fungible/freeze.rs:115]:
```suggestion
	/// Increase the amount which is being frozen for a particular freeze, failing in the case that
```

**Review by franciscoaguirre — APPROVED** [review]

File: substrate/frame/support/src/traits/storage.rs [hunk 6]
```diff
@@ -127,6 +132,66 @@ impl WhitelistedStorageKeys for Tuple {
...
+///
+/// A single ticket corresponding to some particular datum held in storage. This is an opaque
+/// type, but must itself be stored and generally it should be placed alongside whatever data
+/// the ticket was created for.
+///
+/// While not technically a linear type owing to the need for `FullCodec`, *this should be
+/// treated as one*. Don't type to duplicate it, and remember to drop it when you're done with
+/// it.
+#[must_use]
+pub trait Consideration<AccountId>: Member + FullCodec + TypeInfo + MaxEncodedLen {
```
Comment (KiChjang) [inline substrate/frame/support/src/traits/storage.rs:168]: Not sure if the `#[must_use]` attribute put here does anything. I get that the intention is to say that the implementing type must be used in some way, but I don't think the attribute in this position can do so, as the literature didn't document what should happen when placed above a trait item.

Instead, I propose putting it in trait functions that return `Self`, that way we can guarantee that whenever we call the trait function, the result value MUST be used.

Comment (gavofyork) [reply substrate/frame/support/src/traits/storage.rs:168]: [Permalink to the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=94526a438756ad2791faf15800b71a79)

**Comment by gavofyork** [comment]
bot fmt

**Comment by gavofyork** [comment]
bot merge

**Comment by mordamax** [comment]
@gavofyork we've disabled `merge` function since it was used primarily for supporting companions chain. Now github has `automerge` feature which could be used instead

**Pull request merged**